tracing-opentelemetry/src/layer.rs

```diff
@@ -1386,4 +1477,184 @@ mod tests {
             ]))
         );
     }
+
+    #[test]
+    fn tracks_dropped_events() {
+        let tracer = TestTracer(Arc::new(Mutex::new(None)));
+        let subscriber = tracing_subscriber::registry()
+            .with(layer().with_tracer(tracer.clone()).max_events_per_span(4));
+        let subscriber = Arc::new(subscriber);
+
+        tracing::subscriber::with_default(subscriber.clone(), || {
+            let span = tracing::debug_span!("request");
+            let _span_guard = span.enter();
+
+            // we should record all 4 events
+            tracing::info!("event 1");
+            tracing::info!("event 2");
+            tracing::info!("event 3");
+            tracing::info!("event 4");
+
+            let data = span.id().and_then(|id| subscriber.span(&id)).unwrap();
+            let extensions = data.extensions();
+
+            // we shouldn't have recoreded any dropped events
+            assert!(extensions.get::<DroppedOtelEvents>().is_none());
+
+            // we should have all 4 events
+            let otel_data = extensions.get::<OtelData>().unwrap();
+            let otel_events = otel_data.builder.events.as_ref().unwrap();
+            assert_eq!(otel_events.len(), 4);
+
+            drop(extensions);
+            drop(data);
+
+            // record a 5th event, which is over our max of 4
+            tracing::info!("event 5");
+
+            let data = span.id().and_then(|id| subscriber.span(&id)).unwrap();
+            let extensions = data.extensions();
+
+            // since we're over our limit we should start dropping events
+            let dropped_events = extensions.get::<DroppedOtelEvents>().unwrap();
+            assert_eq!(dropped_events.dropped_count, 3);
+
+            let otel_data = extensions.get::<OtelData>().unwrap();
+            let otel_events = otel_data.builder.events.as_ref().unwrap();
+            // dropped 3, pushed one more
+            assert_eq!(otel_events.len(), 2);
+
+            drop(extensions);
+            drop(data);
+
+            // record many events!
+            for i in 6..99 {
+                tracing::info!("event {}", i);
+            }
+
+            let data = span.id().and_then(|id| subscriber.span(&id)).unwrap();
+            let extensions = data.extensions();
+
+            // we should have dropped a ton of events, and recorded that we did
+            let dropped_events = extensions.get::<DroppedOtelEvents>().unwrap();
+            assert_eq!(dropped_events.dropped_count, 96);
+
+            let otel_data = extensions.get::<OtelData>().unwrap();
+            let otel_events = otel_data.builder.events.as_ref().unwrap();
+            assert_eq!(otel_events.len(), 2);
+
+            // even though we pushed a ton of events, the capacity of our buffer
+            // should still be small
+            assert_eq!(otel_events.capacity(), 8);
```

Comment:
if `max_events_per_span` is set, we could use `Vec::reserve` to set this to the exact same size, whenever we resize the vec, right? small optimization, if you are interested, not gonna block merging on it!

Reply:
Good thought! I don't think I'm going to do this though because once we hit our max, then the Vec will already be large enough to store all of our events, so there's no need to resize, and I don't think we want to preemptively reserve enough space, because then a lot of spans would have excess capacity when they don't need it